changelogs/unreleased/gh-5272-bootstrap-strategy-auto.md

```diff
@@ -0,0 +1,8 @@
+## feature/replication
+
+* Introduce new configuration option `bootstrap_strategy`. The default value of
+  this option - "auto" - brings new replica behaviour on replica set bootstrap,
+  recovery and on replication reconfiguration. One of the notable changes is
+  that `replication_connect_quorum` option is not taken into account with
```

Comment:
Isn't it supposed to raise an error when used with non-legacy strategy? At least this is what I see in the RFC.

Reply:
I'm not sure. Asked @Mons and @ochaton.

Reply:
Decided to warn the user, without errors.
This will simplify the upgrade quite a bit.
bootstrap_strategy = "auto" is default, and making all upgrades error, unless `replication_connect_quorum` is removed from the config is bad.
<details><summary>Here's the conversation</summary>
Сергей Петренко (26.12.2022 19:34):
Владимир Перепелица Владислав Грубов, тут во время ревью bootstrap_strategy = "auto" возник вопрос, мы на использование replication_connect_quorum даём ворнинг в лог или кидаем ошибку?

В доке договорились, что кидаем ошибку, но чёто я во время имплементации решил, что не нужно ошибку, чтобы поведение не зависело от значения опции bootstrap_strategy. (Хотя поведение всё равно меняется. RCQ игнорируется, когда bootstrap_strategy = "auto")

Владимир Перепелица (26.12.2022 19:56):
Учитывая скоьлко тут копий сломано, наверное лучше явно обратить внимание и заставить выключить, чтоб не маячило

Сергей Останевич (26.12.2022 19:58):
Upgrade пойдёт успешнее?

Владимир Перепелица (26.12.2022 20:00):
при апгрейде не должно быть опции bootstrap_strategy

Сергей Петренко (26.12.2022 20:10):
> Владимир Перепелица (26.12.2022 20:00): при апгрейде не должно быть опции bootstrap_strategy
в смысле? Мы ж значение по умолчанию делаем "auto"

Владимир Перепелица (26.12.2022 20:11):
Сек, дай подумаю.

Владимир Перепелица (26.12.2022 20:15):
Напрашивается следующее:

если узел/кластер не забутстраплен то:
сочетание явно переданного rcq + bootstrap_strategy — это ошибка.
если явно не передать bootstrap_strategy, но передать rcq, то это варнинг об устаревшей опции (*)
иначе всё ок.

а если узел забутстраплен, то использование rcq — это warning о том, что deprecated.

---
(*) Тут какой момент. Если это сделать ошибкой, то у нас сломается cartridge на простом апгрейде.

Сергей Петренко (26.12.2022 20:22):
> Владимир Перепелица (26.12.2022 20:15): Напрашивается следующее:

если узел/кластер не забутстраплен то:
сочетание явно переданного rcq + bootstrap_strategy — это ошибка.
если явно не передать bootstrap_strategy, но передать rcq, то это варнинг об устаревшей опции (*)
иначе всё ок.

а если узел забутстраплен, то использование rcq — это warning о том, что deprecated.

---
(*) Тут какой момент. Если это сделать ошибкой, то у нас сломается cartridge на простом апгрейде.
Хорошо

Сергей Петренко (26.12.2022 20:25):
Блин. Чёто как-то неконсистентно. Может всегда ворнинг и всё?
Почему
> если узел/кластер не забутстраплен то: сочетание явно переданного rcq + bootstrap_strategy — это ошибка.
А
> если явно не передать bootstrap_strategy, но передать rcq, то это варнинг об устаревшей опции (*)
?

Для тнт внутри эти две ситуации идентичны — box.cfg.bootstrap_strategy == 'auto'
Да и для юзера вроде бы должно быть так же. Раз уж bootstrap_strategy по-умолчанию auto.

Владимир Перепелица (26.12.2022 20:26):
> Сергей Петренко (26.12.2022 20:25): Блин. Чёто как-то неконсистентно. Может всегда ворнинг и всё?
Почему
> если узел/кластер не забутстраплен то: сочетание явно переданного rcq + bootstrap_strategy — это ошибка.
А
> если явно не передать bootstrap_strategy, но передать rcq, то это варнинг об устаревшей опции (*)
?

Для тнт внутри эти две ситуации идентичны — box.cfg.bootstrap_strategy == 'auto'
Да и для юзера вроде бы должно быть так же. Раз уж bootstrap_strategy по-умолчанию auto.
Ок, давай всегда ворнинг. Спилим в 3'ке потом.
</details>